### PR TITLE
fix(Inventory): search Domain with asset entities_id

### PR DESCRIPTION
**Pull request by stonebuzz**
Like ```Software```, ```Domain``` must follow the asset entity

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #number

### CODE REVIEW

File: src/Inventory/Asset/MainAsset.php [hunk 1]
```diff
@@ -710,7 +710,15 @@ public function rulepassed($items_id, $itemtype, $rules_id, $ports_id = 0)
         //handle domains
         if (property_exists($val, 'domains_id')) {
             $domain = new \Domain();
-            if (!$domain->getFromDBByCrit(Sanitizer::sanitize(['name' => $val->domains_id]))) {
+            if (
+                !$domain->getFromDBByCrit(Sanitizer::sanitize(
+                    [
+                        'name' => $val->domains_id,
+                        'is_deleted' => 0,
+                        'entities_id' => $entities_id,
```
Comment (trasher) [inline src/Inventory/Asset/MainAsset.php:718]: Using just `$entitied_id` won't take care of entites inheritance as far as I understand; but it should should'nt it?

Comment (stonebuzz) [reply src/Inventory/Asset/MainAsset.php:718]: I have a doubt, using inheritance could return several results .... =/

**Review by trasher — APPROVED** [review]

File: src/Inventory/Asset/MainAsset.php (hunk 1 above)
Comment (trasher) [reply src/Inventory/Asset/MainAsset.php:718]: I'm almost always lost talking about inheritance :/ Anyways, change proposal will at least fix some cases already, so OK for me.

Comment (cedric-anne) [reply src/Inventory/Asset/MainAsset.php:718]: There is no unicity on `name`+`entities_id`, so using `getFromDBByCrit` could anyway be a problem here.
If we found multiple entries using the entities visibility criteria, it means that there are duplicated entries in the GLPI and it is not the inventory responsibility to fix this.

The following code will permit to attach asset to a domain that has been created in a parent entity, and will also handle presence of multiple domains with the same name in the current entity scope.

```php
$matching_domains = $DB->request([
    'name' => $val->domains_id,
    'is_deleted' => 0,
    'entities_id' => getEntitiesRestrictCriteria($domain->getTable(), '', $entities_id, true),
]);
if ($matching_domains->count() > 0) {
    // Get the first domain, as we assume that a domain should not be declared multiple times in the same entity scope.
    $domain->getFromResultSet($matching_domains->current());
} else {
    $domain->add(...);
}
```

File: src/Inventory/Asset/MainAsset.php [hunk 2]
```diff
@@ -710,7 +710,15 @@ public function rulepassed($items_id, $itemtype, $rules_id, $ports_id = 0)
...
             $domain = new \Domain();
-            if (!$domain->getFromDBByCrit(Sanitizer::sanitize(['name' => $val->domains_id]))) {
+            $matching_domains = $DB->request([
+                'name' => $val->domains_id,
+                'is_deleted' => 0,
+                'entities_id' => getEntitiesRestrictCriteria($domain->getTable(), '', $entities_id, true),
+            ]);
+            if ($matching_domains->count() > 0) {
+                // Get the first domain, as we assume that a domain should not be declared multiple times in the same entity scope.
+                $domain->getFromResultSet($matching_domains->current());
```
Comment (cedric-anne) [inline src/Inventory/Asset/MainAsset.php:720]:
```suggestion
            $matching_domains = $DB->request([
                'FROM' => $domain->getTable(),
                'WHERE' => [
                    'name' => $val->domains_id,
                    'is_deleted' => 0,
                    'entities_id' => getEntitiesRestrictCriteria($domain->getTable(), '', $entities_id, true),
                ],
                'LIMIT' => 1, // Get the first domain, as we assume that a domain should not be declared multiple times in the same entity scope
            ]);
            if ($matching_domains->count() > 0) {
                $domain->getFromResultSet($matching_domains->current());
```

File: src/Inventory/Asset/MainAsset.php [hunk 3]
```diff
@@ -710,7 +710,18 @@ public function rulepassed($items_id, $itemtype, $rules_id, $ports_id = 0)
...
         if (property_exists($val, 'domains_id')) {
             $domain = new \Domain();
-            if (!$domain->getFromDBByCrit(Sanitizer::sanitize(['name' => $val->domains_id]))) {
+            $matching_domains = $DB->request([
+                'FROM' => $domain->getTable(),
+                'WHERE' => [
+                    'name' => $val->domains_id,
+                    'is_deleted' => 0,
+                    'entities_id' => getEntitiesRestrictCriteria($domain->getTable(), '', $entities_id, true),
+                ],
```
Comment (cedric-anne) [inline src/Inventory/Asset/MainAsset.php:719]:
```suggestion
                ] + getEntitiesRestrictCriteria($domain->getTable(), '', $entities_id, true),
```

File: src/Inventory/Asset/MainAsset.php [hunk 4]
```diff
@@ -710,7 +710,17 @@ public function rulepassed($items_id, $itemtype, $rules_id, $ports_id = 0)
         //handle domains
         if (property_exists($val, 'domains_id')) {
             $domain = new \Domain();
-            if (!$domain->getFromDBByCrit(Sanitizer::sanitize(['name' => $val->domains_id]))) {
+            $matching_domains = $DB->request([
+                'FROM' => $domain->getTable(),
+                'WHERE' => [
+                    'name' => $val->domains_id,
```
Comment (cedric-anne) [inline src/Inventory/Asset/MainAsset.php:716]:
```suggestion
                    'name' => Sanitizer::sanitize($val->domains_id),
```

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**